eth/handler.go

```diff
@@ -473,6 +461,28 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 		}
 		return p.SendBlockBodiesRLP(bodies)
 
+	case p.version >= eth62 && msg.Code == BlockBodiesMsg:
+		// A batch of block bodies arrived to one of our previous requests
```

Comment:
We only request a single message of a specific kind from a peer at any time. One body fetch, one header, one receipt, one node data. The reply is assumed to be for that request. There's of course nothing stopping a peer from sending out of bound data, but that will just get ignored.

Note, there's one exception. There's a very tiny probability that the fetcher and downloader request blocks the same time from the same peer when transitioning from "doing a big sync" to "staying in sync", but the fetcher will explicitly cherry pick it's own requests and leave all others to the downloader, so this does not bother them.